mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp

```diff
@@ -55,6 +61,85 @@ void createAttnTuningRangeBF(TuningParamSet *newSpace, AttentionOp attnOp,
   }
 }
 
+SmallVector<int64_t>
+computeOptimalSplitKFactors(RockGemmWrapperInterface gemmOp,
+                            int32_t gemmMPerBlock, int32_t gemmNPerBlock,
+                            int32_t gemmKPerBlock, int32_t kPack) {
+  auto info = PopulateParamsInfo::fromOp(gemmOp);
+  SmallVector<int64_t> splitKValues = {1};
+  if (!info.gemmAType.isF32()) {
+    return splitKValues;
+  }
+
+  // TODO[split-K]: remove after integrating split-K into MIGraphX
+  auto func = llvm::cast<func::FuncOp>(gemmOp->getParentOp());
+  if (!func->hasAttr(rock::EnableSpliKForTuningAttr::getMnemonic())) {
+    return splitKValues;
+  }
+
+  if (!gemmOp.getNumCU().has_value()) {
+    return splitKValues;
+  }
+  const double numCUs = gemmOp.getNumCU().value();
+
+  const InitParams params{gemmMPerBlock, gemmNPerBlock, gemmKPerBlock};
+  const GemmSize gemmSize =
+      calculatePaddedGemmSize(params, info.gemmSize, kPack);
+  const double numMTiles = gemmSize.m / gemmMPerBlock;
+  const double numNTiles = gemmSize.n / gemmNPerBlock;
+
+  auto computeImbalance = [&](int32_t splitKFactor) -> double {
+    const double totalNumWorkGroups =
+        gemmSize.g * numMTiles * numNTiles * splitKFactor;
+    const double maxWorkGroupsPerCU = std::ceil(totalNumWorkGroups / numCUs);
+    // imbalances = max. CU work / average work per CU
+    return (maxWorkGroupsPerCU * numCUs) / totalNumWorkGroups;
+  };
+
+  const auto dataPrallelGemmImbalance = computeImbalance(1);
+  constexpr double imbalaceThreshold = 1.20;
+  if (dataPrallelGemmImbalance < imbalaceThreshold) {
+    return splitKValues;
+  }
+
+  struct LocalData {
+    int64_t splitKValue = 0;
+    double workImbalance = 0.0;
+  };
+  SmallVector<LocalData> factors;
+  constexpr double minGain = 1.30;
+  // There are cases where perfect load balancing can be achieved with very
+  // high splitK values. However, experiments show that performance
+  // can considerably drop in such cases. Currently, we limit the `upperBound`
+  // on purpose because the current heuristics does not consider the overheads
+  // resulting from reducing partial solution along the split dimension.
+  // This needs to be improved in the future.
+  constexpr int32_t upperBound = 32;
+  for (int32_t splitKFactor = 2; splitKFactor < upperBound; ++splitKFactor) {
+    const double imbalance = computeImbalance(splitKFactor);
+    const auto gain = dataPrallelGemmImbalance / imbalance;
+    if (gain > minGain) {
+      factors.emplace_back(LocalData{splitKFactor, imbalance});
+    }
+  }
+
+  if (factors.empty()) {
+    return splitKValues;
+  }
+
+  llvm::sort(factors, [](LocalData &a, LocalData &b) {
+    return a.workImbalance < b.workImbalance;
+  });
+
+  const size_t maxVariants = std::min(static_cast<size_t>(6), factors.size());
+  llvm::ArrayRef<LocalData> view(factors.data(), maxVariants);
+  llvm::for_each(view, [&](const LocalData &item) {
+    splitKValues.push_back(item.splitKValue);
+  });
+
+  return splitKValues;
+}
```

Comment:
Since we're getting codecov warnings here that kinda make sense, can you add to or adapt `mlir/test/rocmlir-gen/emit-tuning-space.mlir` to test this split-K sorting heuristic?

Reply:
Yep. I added an additional test, which actually captured a small bug. Thanks!